helix-core/src/main/java/org/apache/helix/controller/dataproviders/BaseControllerDataProvider.java

```diff
@@ -744,30 +747,31 @@ private void updateOfflineInstanceHistory(HelixDataAccessor accessor) {
     _updateInstanceOfflineTime = false;
   }
 
-  private void updateDisabledInstances() {
+  private void updateDisabledInstances(Collection<InstanceConfig> instanceConfigs,
+      ClusterConfig clusterConfig) {
     // Move the calculating disabled instances to refresh
     _disabledInstanceForPartitionMap.clear();
     _disabledInstanceSet.clear();
-    for (InstanceConfig config : _instanceConfigCache.getPropertyMap().values()) {
+    for (InstanceConfig config : instanceConfigs) {
       Map<String, List<String>> disabledPartitionMap = config.getDisabledPartitionsMap();
       if (!config.getInstanceEnabled()) {
         _disabledInstanceSet.add(config.getInstanceName());
       }
       for (String resource : disabledPartitionMap.keySet()) {
         if (!_disabledInstanceForPartitionMap.containsKey(resource)) {
-          _disabledInstanceForPartitionMap.put(resource, new HashMap<String, Set<String>>());
+          _disabledInstanceForPartitionMap.put(resource, new HashMap<>());
         }
         for (String partition : disabledPartitionMap.get(resource)) {
           if (!_disabledInstanceForPartitionMap.get(resource).containsKey(partition)) {
-            _disabledInstanceForPartitionMap.get(resource).put(partition, new HashSet<String>());
+            _disabledInstanceForPartitionMap.get(resource).put(partition, new HashSet<>());
           }
           _disabledInstanceForPartitionMap.get(resource).get(partition)
               .add(config.getInstanceName());
         }
       }
     }
-    if (_clusterConfig != null && _clusterConfig.getDisabledInstances() != null) {
-      _disabledInstanceSet.addAll(_clusterConfig.getDisabledInstances().keySet());
+    if (clusterConfig != null && clusterConfig.getDisabledInstances() != null) {
```

Comment:
again, is `clusterConfig` passed in would always to the same as `_clusterConfig`?

Reply:
Please refer to the previous comment.